core/src/main/java/com/segment/analytics/kotlin/core/utilities/OkHttpURLConnection.kt

```diff
@@ -0,0 +1,393 @@
+                builder.post(body)
+            }
+            "PUT" -> {
+                val body = requestBodyBuffer?.readByteArray()?.toRequestBody(
+                    getRequestProperty("Content-Type")?.toMediaType() ?: "text/plain".toMediaType()
+                ) ?: "".toRequestBody("text/plain".toMediaType())
+                builder.put(body)
+            }
+            "DELETE" -> builder.delete()
+            "HEAD" -> builder.head()
+            else -> throw ProtocolException("Unknown method: $_requestMethod")
+        }
+
+        return builder.build()
+    }
+
+    @Throws(IOException::class)
+    override fun getInputStream(): InputStream {
+        if (!connected) connect()
+        return response?.body?.byteStream() ?: throw IOException("No response body")
+    }
+
+    @Throws(IOException::class)
+    override fun getErrorStream(): InputStream? {
+        if (!connected) connect()
+        return if (responseCode >= 400) {
+            response?.body?.byteStream()
+        } else null
+    }
+
+    @Throws(IOException::class)
+    override fun getOutputStream(): OutputStream {
+        // Automatically switch to POST if method is GET (mimics standard HttpURLConnection behavior)
+        if (_requestMethod == "GET") {
+            _requestMethod = "POST"
+        }
+        
+        if (requestBodyBuffer == null) {
+            requestBodyBuffer = Buffer()
+        }
+        return requestBodyBuffer!!.outputStream()
+    }
+
+    override fun getHeaderField(name: String?): String? {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return null }
+        }
+        return response?.header(name ?: return null)
+    }
+
+    override fun getHeaderFields(): MutableMap<String, MutableList<String>> {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return mutableMapOf() }
+        }
+        return response?.headers?.toMultimap()?.mapValues { it.value.toMutableList() }?.toMutableMap()
+            ?: mutableMapOf()
+    }
+
+    override fun getResponseCode(): Int {
+        if (!connected) connect()
+        return responseCode
+    }
+
+    override fun getResponseMessage(): String {
+        if (!connected) connect()
+        return responseMessage ?: ""
+    }
+
+    /**
+     * Returns the protocol used for this connection (e.g., "h2" for HTTP/2, "http/1.1" for HTTP/1.1).
+     * Returns null if the connection hasn't been established yet.
+     */
+    fun getProtocol(): String? {
+        return response?.protocol?.toString()
+    }
+
+    // Override ALL URLConnection methods to prevent bypassing OkHttp
+    override fun getURL(): URL = url
+
+    override fun getHeaderFieldKey(n: Int): String? {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return null }
+        }
+        val headers = response?.headers ?: return null
+        return if (n in 0 until headers.size) headers.name(n) else null
```

Comment:
This is confusing to me. If headers = null, will the call in the if to headers.size work and not throw a NPE?

Reply:
certainly not the most verbose way of doing it 😅

Reply:
no. line 165 will already return null if `response?.headers` is null